src/software/simulation/threaded_simulator.cpp

```diff
@@ -0,0 +1,87 @@
+
+ThreadedSimulator::ThreadedSimulator(const Field &field) : simulator(field), simulation_thread_started(false), stopping_simulation(false) {}
+
+ThreadedSimulator::~ThreadedSimulator() {
+    stopSimulation();
+}
+
+void ThreadedSimulator::registerOnSSLWrapperPacketReadyCallback(const std::function<void (SSL_WrapperPacket)>& callback) {
+    ssl_wrapper_packet_callbacks.emplace_back(callback);
+}
+
+void ThreadedSimulator::startSimulation() {
+    // Start the thread to do the simulation in the background
+    if(!simulation_thread_started.load()) {
+        simulation_thread_started = true;
+        simulation_thread = std::thread(&ThreadedSimulator::runSimulationLoop, this);
+    }
+}
+
+void ThreadedSimulator::stopSimulation() {
+    if (simulation_thread_started.load())
+    {
+        simulation_thread_started = false;
+
+        // Set this flag so the simulation_thread knows to end
+        stopping_simulation = true;
+
+        simulation_thread.join();
+
+        // Reset the flag in case we want to re-start the simulation
+        stopping_simulation = false;
+    }
+}
+
+void ThreadedSimulator::setBallState(const BallState &ball_state) {
+    std::scoped_lock lock(simulator_mutex);
+    simulator.setBallState(ball_state);
+}
+
+void ThreadedSimulator::removeBall() {
+    std::scoped_lock lock(simulator_mutex);
+    simulator.removeBall();
+}
+
+void ThreadedSimulator::addYellowRobots(const std::vector<RobotStateWithId> &robots) {
+    std::scoped_lock lock(simulator_mutex);
+    simulator.addYellowRobots(robots);
+}
+
+void ThreadedSimulator::addBlueRobots(const std::vector<RobotStateWithId> &robots) {
+    std::scoped_lock lock(simulator_mutex);
+    simulator.addBlueRobots(robots);
+}
+
+void ThreadedSimulator::setYellowRobotPrimitives(ConstPrimitiveVectorPtr primitives) {
+    std::scoped_lock lock(simulator_mutex);
+    simulator.setYellowRobotPrimitives(primitives);
+}
+
+void ThreadedSimulator::setBlueRobotPrimitives(ConstPrimitiveVectorPtr primitives) {
+    std::scoped_lock lock(simulator_mutex);
+    simulator.setBlueRobotPrimitives(primitives);
+}
+
+void ThreadedSimulator::runSimulationLoop() {
+    Duration time_step = Duration::fromSeconds(TIME_STEP_SECONDS);
+
+    while(!stopping_simulation.load()) {
+        auto simulation_step_start_time = std::chrono::steady_clock::now();
+
+        simulator_mutex.lock();
+        simulator.stepSimulation(time_step);
+        auto ssl_wrapper_packet_ptr = simulator.getSSLWrapperPacket();
+        simulator_mutex.unlock();
+
+        assert(ssl_wrapper_packet_ptr);
+        SSL_WrapperPacket ssl_wrapper_packet = *(ssl_wrapper_packet_ptr.release());
+
+        for(const auto& callback : ssl_wrapper_packet_callbacks) {
+            callback(ssl_wrapper_packet);
+        }
+
+        auto simulation_step_end_time = simulation_step_start_time + std::chrono::microseconds(static_cast<unsigned int>(TIME_STEP_SECONDS * MICROSECONDS_PER_SECOND));
```

Comment:
what happens if this time is in the past? It feels like we should log a warning

Reply:
it means we'd be running slower than real-time, which will probably show up visually? Is it worth a warning log? Assuming this gets run in a standalone executable, we may not be able to see the console anyway.

Reply:
I think it probably makes more sense to keep some idea of how fast relative to realtime the simulator is running, that we can publish somehow and display in the visualizer.

Reply:
Fair enough, kind of like an FPS value. It wouldn't necessarily be a log so I've created #1491 so deal with this separately, since it's somewhat out of scope for this PR.